src/Memory/InputQueue.ts

```diff
@@ -6,117 +6,149 @@ import * as BB from 'botbuilder'
+
+    /**
+     * 
+     * @param inProcessMessageState Store for mutex
+     * @param activity Input that needs to be handled
+     * @param callback To be called when input is ready to be handled
+     */
+    public static async AddInput(inProcessMessageState: InProcessMessageState, activity: BB.Activity, callback: Function): Promise<void> {
         if (!activity.id) {
+            CLDebug.Error("InputQueue: Activity has no activityId")
             return
         }
 
         // Add to queue
-        await InputQueue.InputQueueAdd(activity.id, callback)
+        await InputQueue.InputQueueAdd(activity.conversation.id, activity.id, callback)
 
         // Process queue
-        await InputQueue.InputQueueProcess(inProcessMessageState)
+        await InputQueue.InputQueueProcessNext(inProcessMessageState, activity.conversation.id)
     }
 
-    // Add message to queue
-    private static async InputQueueAdd(activityId: string, callback: Function): Promise<void> {
+    // Add input to queue
+    private static async InputQueueAdd(conversationId: string, activityId: string, callback: Function): Promise<void> {
         const now = new Date().getTime()
         const queuedInput: QueuedInput = {
+            conversationId,
             activityId,
             timestamp: now,
             callback,
         }
 
-        this.messageQueue.push(queuedInput)
-        CLDebug.Log(`ADD QUEUE: ${activityId} ${this.messageQueue.length}`, DebugType.MessageQueue)
+        if (!this.inputQueues[conversationId]) {
+            this.inputQueues[conversationId] = []
+        }
+        this.inputQueues[conversationId].push(queuedInput)
+        this.log(`ADD QUEUE`, conversationId, queuedInput.activityId)
     }
 
-    // Attempt to process next message in the queue
-    private static async InputQueueProcess(inProcessMessageState: InProcessMessageState): Promise<void> {
+    private static hasExpired(queuedInput: QueuedInput): boolean {
         const now = new Date().getTime()
-        const messageProcessing = await inProcessMessageState.get<QueuedInput>()
+        const age = now - queuedInput.timestamp
+        return (age > MESSAGE_TIMEOUT)
+    }
+
+    // Process next input
+    private static async InputQueueProcessNext(inProcessMessageState: InProcessMessageState, conversationId: string): Promise<void> {
 
-        // Is a message being processed
-        if (messageProcessing) {
+        // Get current input being processed (mutex)
+        let inputInProcess = await inProcessMessageState.get<QueuedInput>()
 
-            // Remove if it's been expired
-            const age = now - messageProcessing.timestamp
+        // If input is being processed (mutex is set), check if it has expired
+        if (inputInProcess) {
+            if (this.hasExpired(inputInProcess)) {
+                // Item in mutex has expired
+                this.log(`EXPIRED U`, conversationId, inputInProcess.activityId)
```

Comment:
What is the `U` for?

Reply:
I see the other one with `Q`, whatever it stands for, write it out.

Reply:
Just for logging.  Keeping the num or chars constant so logs line up in console - much easier to read and see what's going on.